File: 1028-interval-list-intersections/interval-list-intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        int i=0;
        int j=0;
        vector<vector<int>> ans;
        while(i<firstList.size() && j<secondList.size()){
            int t1 = max(firstList[i][0],secondList[j][0]);
            int t2 = min(firstList[i][1],secondList[j][1]);
            if(t1<=t2){//for intersection to exist this should be valid
                ans.push_back({t1,t2});
            }
            if(firstList[i][1] < secondList[j][1]) i++;//traversing next elements
            else j++;
        }
        return ans;
        
    }
};

File: 1028-interval-list-intersections/interval-list-intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
        List<int[]> ans = new ArrayList<>();
        int i = 0;
        int n = firstList.length;
        int j = 0;
        int m = secondList.length;
        while(i<n && j<m){
            if(firstList[i][0]<=secondList[j][1] && firstList[i][1]>=secondList[j][0])
            {
                ans.add(new int[]{Math.max(firstList[i][0],secondList[j][0]),Math.min(firstList[i][1],secondList[j][1])});
            }
            if(firstList[i][1]<secondList[j][1])
            {
                i++;
            }
            else j++;
        }

       int[][] result = ans.toArray(new int[ans.size()][]);
       return result;
        
    }
}